femtools/Field_derivatives.F90

```diff
@@ -42,16 +42,280 @@ module field_derivatives
       domain_is_2d, patch_type, get_patch_ele, get_patch_node, get_quadratic_fit_qf, curl, &
-      get_quadratic_fit_eqf, div, u_dot_nabla, get_cubic_fit_cf, differentiate_field_lumped
+      get_quadratic_fit_eqf, div, u_dot_nabla, get_cubic_fit_cf, differentiate_field_lumped, &
+      dg_ele_grad_at_quad, dg_ele_grad
 
     public :: compute_hessian_qf, compute_hessian_eqf, compute_hessian_var
     
     contains
 
+    function dg_ele_grad_scalar(field, ele, X, bc_value, bc_type) result (loc_grad)
+      ! Return the grad of field at the quadrature points of
+      ! ele_number. dn is the transformed element gradient. 
+      ! including interface terms for dg discretisations based upon
+      ! N_i N_j grad_u = N_i delta u_h - 
+      !                  ({N_i} (u_h^-n^- + u_h^+n^+)) on internal faces -
+      !                  (N_i (u_h - u_b) n) on weak dirichlet boundaries
+      ! where: {x} = average of x over face
+      !        u_h = value of u in element
+      !        u_b = dirichlet boundary value
+      ! (see Bassi et. al. 2005 - Discontinuous Galerkin solution of the Reynolds-averaged
+      ! Navier–Stokes and k–x turbulence model equations, pg. 517
+      type(scalar_field),intent(in) :: field
+      integer, intent(in) :: ele 
+      type(vector_field), intent(in) :: X
+      type(scalar_field), intent(in), optional :: bc_value
+      integer, dimension(:), intent(in), optional :: bc_type   
+      real, dimension(mesh_dim(field), ele_loc(field, ele)) :: loc_grad
+
+      ! variables for interior integral
+      type(element_type), pointer :: shape
+      real, dimension(ele_loc(field, ele), ele_ngi(field, ele), mesh_dim(field)) :: dshape
+      real, dimension(ele_ngi(field, ele)) :: detwei
+      real, dimension(mesh_dim(field), ele_ngi(field, ele)) :: grad_h_gi
+
+      ! variables for surface integral
+      integer :: ni, ele_2, face, face_2, i
+      integer, dimension(:), pointer :: neigh
+
+      ! inverse mass
+      real, dimension(ele_loc(field, ele), ele_loc(field, ele)) :: inv_mass
+
+      ! assert that we have either both bc_value and bc_type, or neither
+      assert((present(bc_value) .and. present(bc_type)) .or. (.not. present(bc_value) .and. .not. present(bc_type)))
+
+      ! we can only compute for the L1 halo as we need values from adjacent elements
+      assert( element_owned(field, ele) .or. element_neighbour_owned(field, ele) )
+
+      shape => ele_shape(field, ele) 
+      call transform_to_physical(X, ele, shape, dshape, detwei)
+
+      ! Calculate grad within the element
+      grad_h_gi = ele_grad_at_quad(field, ele, dshape)
+
+      ! Assemble interior contributions to rhs
+      loc_grad = shape_vector_rhs(shape, grad_h_gi, detwei)
+
+      ! Interface integrals
+      neigh=>ele_neigh(field, ele)
+      do ni=1,size(neigh)
+        ! Find the relevant faces.
+        ele_2 = neigh(ni)
+        face = ele_face(field, ele, ele_2)
```

Comment:
This is very inefficient: it loops again over all facets of ele to find the neighbouring ele_2. Instead you can use a faces => ele_faces(field, ele) before the loop, and just face = faces(ni) here.